### ISSUE TITLE
TypeError: allowed.indexOf is not a function using IO.js v3.3.1

### ISSUE THREAD

**Issue by zrecommerce** [Swaagie/minimize#63]
```
/home/aalbino/Projects/xtal-commerce/node_modules/gulp-minify-html/node_modules/minimize/lib/helpers.js:132
      ? allowed === '*' || ~allowed.indexOf(name)
                                    ^

TypeError: allowed.indexOf is not a function
```

Looks like this:
```
    //
    // Is the attribute allowed on the HTML element? If so, allow special
    // treatment. If not, then just return the full attribute and its value.
    //
    allowed = list.attributes[key];
    allowed = allowed
      ? allowed === '*' || ~allowed.indexOf(name)
      : ~key.indexOf('data-');
```

...should in fact be:

```
    //
    // Is the attribute allowed on the HTML element? If so, allow special
    // treatment. If not, then just return the full attribute and its value.
    //
    allowed = String(list.attributes[key]); // <-- Should be a String to provide indexOf() method.
    allowed = allowed
      ? allowed === '*' || ~allowed.indexOf(name)
      : ~key.indexOf('data-');
```

**Comment by Swaagie** [comment]
This is a legit issue, but the source is probably due to `list.attributes[key]`returning an `undefined` value, calling string on that will create a string `'undefined'`. Thats not what we want, rather, I'll do something like this `allowed = list.attributes[key] || '*';` as allowed is either an array or asterix. The cause is an unknown element type that is not prevalent in the list here: https://github.com/Swaagie/minimize/blob/master/lib/list.js#L46. Out of curiosity what was the element or part of the code that caused this error?

**Comment by Swaagie** [comment]
Actually judging from the comment just above it `allowed` can be falsey, hence I cannot default it to an asterix. However, the question remains what value does `allowed` have? Both arrays and strings have `indexOf`, an undefined value would return a different error, so the remaining suspects are Boolean, Object or Buffer?

**Comment by Swaagie** [comment]
Finally all tests are running fine for me in io.js 3.3.1, could you provide a test case that fails?

**Comment by zrecommerce** [comment]
No problem. I am able to replicate the issue locally, so I should have no trouble finding what the unknown value actually is. Stay tuned...

**Comment by zrecommerce** [comment]
Ok, a quick run with some console output sheds a bit of light...

It is a function, the console.log() call shows it is a 'function Object()'

```
'allowed' is of type:  function , dump:  function Object() { [native code] }
/home/aalbino/Projects/xtal-commerce/node_modules/gulp-minify-html/node_modules/minimize/lib/helpers.js:133
      ? allowed === '*' || ~allowed.indexOf(name)
                                    ^

TypeError: allowed.indexOf is not a function
    at /home/aalbino/Projects/xtal-commerce/node_modules/gulp-minify-html/node_modules/minimize/lib/helpers.js:133:37
```

**Comment by Swaagie** [comment]
Are we talking about https://elements.polymer-project.org/ polymer elements here? If so how is the HTML looking? since attributes is a predefined list only containing array's and strings, I'm still unsure how it suddenly would become an object.

**Comment by zrecommerce** [comment]
Yes. I have a polymer front-end that is being run through vulcanize. It appears the gulp-minify-html npm module uses minimize.

**Comment by Swaagie** [comment]
True, no doubts about that, but could you show me some HTML/polymer code
that might cause the issue? Or if your code is proprietary, could you give
me a similar example that triggers the same error.
Op 14 okt. 2015 5:24 PM schreef "Alex Albino" <notifications@github.com>:

**Comment by zrecommerce** [comment]
It is an open-source app. https://github.com/xtalio/xtal-commerce

**Comment by Swaagie** [comment]
ok thanks, I hadn't noticed that before, so which script of the app should I be running to see the error.

**Comment by zrecommerce** [comment]
After installing npm dependencies, simply run gulp from the base directory.

**Comment by Swaagie** [comment]
After doing that I get several `{ [Error: spawn sass ENOENT] ... `

**Comment by zrecommerce** [comment]
probably needs the sass ruby gem installed

**Comment by Swaagie** [comment]
That makes sense, it now dies on ```Error: Uncaught, unspecified "error" event. (gulp-useref: no such file or directory '/Users/mswaagman/projects/xtal-commerce/{.tmp,app,dist}/bower_components/webcomponentsjs/webcomponents.js')```

**Comment by Swaagie** [comment]
there we go I can replicate it now, I'll investigate it a bit

**Comment by Swaagie** [comment]
fixed with https://github.com/Swaagie/minimize/commit/dd2340870de3503747665a2c49bebd6ab46f49b7, they `key` value was `constructor` and that property on `attributes.list` is the `constructor` method. Hehe weirdest edge case I've seen thus far, but glad we could figure this out. Released a new version 1.7.3 with the fix, so in theory it should work after a reinstall.

**Comment by zrecommerce** [comment]
wow awesome! thanks :+1:
